File: Python/11003_Boxes.py
from sys import stdin

P = [0] * 1000
L = [0] * 1000

dp = [[0] * 6003 for _ in range(1001)]

def solve(N, ML):
    for n in range(N, -1, -1):
        for ml in range(0, ML + 1):
            if n == N:
                dp[n][ml] = 0
            else:
                if ml - P[n] >= 0:
                    dp[n][ml] = max(1 + dp[n + 1][min(ml - P[n], L[n])], dp[n + 1][ml])
                else:
                    dp[n][ml] = dp[n + 1][ml]
    return dp[0][6002]

def main():
    
    N = int(stdin.readline())
    while N != 0:
        for n in range(N):
            p, l = map(int, input().split())
            P[n] = p
            L[n] = l
        ans = solve(N, 6002)
        print(ans)
        N = int(stdin.readline())
        

main()

File: UVAOnlineJudge/Cpp/11003_Boxes.cpp
#include <iostream>

//it can be improved reducing the tab space, but i dont want to think too much about it.

using namespace std;

int P[1000];
int L[1000];



int dp[1001][6002];


int solve(int N, int ML)
{
    for (int n = N; n >= 0; n--)
    {
        for (int ml = 0; ml <= ML; ml++)
        {
            if (n == N)
            {
                dp[n][ml] = 0;
            }else{
                if (ml-P[n] >= 0)
                {
                    dp[n][ml] = max(1+dp[n+1][min(ml-P[n], L[n])], dp[n+1][ml]);
                }else{
                    dp[n][ml] = dp[n+1][ml];
                }
            }
        }
    }
    /*
    for (int n = N; n >= 0; n--)
    {
        for (int ml = 0; ml <= ML; ml++)
        {
            cout<<dp[n][ml]<<",";
        }
        cout<<endl;
    }*/


    return dp[0][6001];
}

int main()
{
    int N;
    while (cin>>N && N != 0)
    {
        for (int n = 0; n<N; n++)
        {
            cin>>P[n]>>L[n];
        }
        int ans = solve(N, 6002);
        cout<<ans<<endl;
    }

    return 0;
}